Big refactor, better multi-server support, DB container for dev.

cpu opened the pull request:
This is a big PR that refactors the old `caa-lookups` based code initially imported into this repo. It reorganizes things slightly by pulling most of the code out of the `dnslol` command into a `dnslol` package that `cmd/dnslol/main.go` can reference. In the future this will make writing unit tests easier.

The changes in this branch lay the ground work for comparing results between nameservers. The old `caa-lookups` code would only round-robin between configured servers. Now it is assumed queries should be sent to all configured servers for comparison. The output & Prometheus stats are updated to ensure its possible to differentiate results between servers.

In preparation for storing the results in a database instead of printing them to standard out for processing with shell based tools this branch includes a small `docker-compose.yml` file that can be used to stand up a MariaDB instance. Right now nothing is using this DB instance but that's where I'm heading next development-wise.

Commit:
Add MariaDB container docker-compose

Commit:
Refactor WIP

Commit:
Rename lookups.go -> main.go

Commit:
Merge remote-tracking branch 'le/master' into cpu-db-support

Commit:
dnslol: incorporate parseServers PR

Commit:
Fix -metricsAddr arg in example

Commit:
dnslol: check ulimit at startup

Commit:
Better multi-server support

Commit:
Better godoc

Commit:
Fix formatting directives

Commit:
Clarify DB support isn't ready yet

Commit:
Remove extra error prefix

jsha commented on dnslol/dnslol.go:
```diff
@@ -0,0 +1,310 @@
+		labels := strings.Split(name, ".")
+		for i := 0; i < len(labels); i++ {
+			queries = append(queries,
+				queryPerServer(strings.Join(labels[i:], "."), dns.TypeCAA)...)
```
This seems to run into a problem with the `query` abstraction. Really what we want for CAA is a single result for `www.example.com`, even though that involved 3 lookups under the hood. I believe this approach will log multiple results, one for `www.example.com`, one for `example.com`, and one for `com`. My recommendation: Just rip out CAA support until we have time to add it back in, since we're not that interested in CAA testing with this tool right now.

cpu replied:
It indeed does log multiple results for each label of the domain (example output is in the PR desc). That seemed reasonable when I assumed it was going to be putting those results into a database soon instead and it would be easier to reconstruct the results from multiple labels for one domain.

Ripping it out for now is fine with me :+1:

cpu replied:
Removed in 7c035a4

jsha commented on dnslol/dnslol.go:
```diff
@@ -0,0 +1,310 @@
+	for _, answer := range in.Answer {
+		// We additionally check that CAA records don't have a case mismatch
+		//
+		// TODO(@cpu): This smells like something that should be behind
```
Let's rip out the CAA case detection stuff. It's not needed anymore.

cpu replied:
Removed in 7c035a4

jsha commented on dnslol/dnslol.go:
```diff
@@ -0,0 +1,310 @@
+// other than RcodeSuccess, a false bool and a nil error or an empty rcode
+// string, a false bool and a non-nil error. In all cases the queryTimes latency
+// stat is updated for the server and query type performed.
+func (e Experiment) queryOne(dnsClient *dns.Client, q query) (string, bool, error) {
```
Having three return types here seems to make the code harder to understand - as evidenced by the fact that you need to point out the middle value is a 'success bool' at each return site. What about just returning a single `error` value? If the error is non-`nil`, that can be considered a success. If you want to distinguish various non-SUCESS Rcodes from network-layer errors, you could define a new `RcodeError` type that callers could use to tell them apart.

cpu replied:
The `RcodeError` type sounds like a better approach, I agree this is a bit jank :laughing:

cpu replied:
Fixed in dd9b3f0 - a separate err type wasn't required.

jsha commented on docker-compose.yml:
```diff
@@ -0,0 +1,9 @@
+db:
+  image: mariadb:10.3
+  ports:
+    - "3333:3306"
```
Better to not publish the port since that exposes the port on the host's external interfaces too. Instead, if we assign a specific IP to this container (see Boulder's docker-compose), then the host can connect to it on a reliable address, but won't further export the port.

cpu replied:
Boulder's docker compose is setting up separate networks - is that what you're suggesting for here? It seemed like overkill

jsha replied:
Yeah, though we don't need multiple ones. You could accomplish the same without specifically configuring a subnet, by just connecting to whatever IP docker happens to assign to the instance. But you'd probably have to do something like `docker inspect boulder_bmysql_1 '{{NetworkSettings.IPv4Address}}'` to find out the IP address when connecting to it from the host.

cpu replied:
Yeah, that's the awkward part :-/ I wish there was a way to just expose the port on localhost of the host. Without a better idea I'll just update the docs to find the docker IP and remove the ports mapping.

jsha replied:
The subnet approach seems to be recommend on stackoverflow: https://stackoverflow.com/a/40573218/363869

cpu replied:
Fixed in d8f3212 - thanks!

jsha commented on cmd/dnslol/main.go:
```diff
@@ -0,0 +1,194 @@
+		"servers",
+		"127.0.0.1:53",
+		"Comma-separated list of DNS servers")
+	roundRobinFlag = flag.Bool(
```
We can ditch the round robin stuff. I think I had some idea that instead of one beefy DNS server, I'd set up eight or so smaller ones, to be able to scale out more easily. But I've never actually done that.

Also, in general if we did go down that route, we'd want two pools: One for the control group, and one for the experiment group. So maintaining a roundrobin strategy separate from the comparison strategy probably isn't fruitful.

cpu replied:
Makes sense :+1: I'll remove it. I can't think of a good use case for it either.

cpu replied:
Removed in 7adf1d6 - without the robin robin stuff the selector interface/abstraction wasn't useful so I removed it too.

Commit:
Remove CAA lookup support pending more thought/design.

We're not presently interested in looking at CAA testing with this tool
and the semantics of tree-climbing make it a little bit difficult to
integrate into the codebase as-is. Rip out CAA support for now, we can
always add it back when we have more time and a specific use-case in
mind.

Commit:
Docker: use static IP for DB container.

Commit:
Simplify queryOne

Commit:
Fix extra output space

Commit:
Remove selectors/round-robin server support.

Commit:
remove duplicate servers check

Commit:
Use .2 for DB IP, fix port #.
